lib/Arch/SPARC32_C.cpp

```diff
@@ -0,0 +1,347 @@
+    }
+
+    case llvm::Type::DoubleTyID: {
+      ret_values.emplace_back();
+      auto &value_declaration = ret_values.back();
+      value_declaration.reg = arch->RegisterByName("d0");
+      value_declaration.type = ret_type;
+      return llvm::Error::success();
+    }
+
+    // Try to split the composite type over registers, and fall back on RVO
+    // if it's not possible.
+    case llvm::GetFixedVectorTypeId():
+    case llvm::Type::ArrayTyID:
+    case llvm::Type::StructTyID: {
+      AllocationState alloc_ret(return_register_constraints, arch, this);
+      alloc_ret.config.type_splitter = IntegerTypeSplitter;
+      auto mapping = alloc_ret.TryRegisterAllocate(*ret_type);
+
+      // There is a valid split over registers, so add the mapping
+      if (mapping) {
+        return alloc_ret.CoalescePacking(mapping.getValue(), ret_values);
+
+      // Composite type splitting; Unlike with x86, LLVM doesn't naturally
+      // perform RVO on large structures returned by value from bitcode.
+      } else {
+        break;
+      }
+    }
+
+    default:
+      break;
+  }
+
+  return llvm::createStringError(
+      std::errc::invalid_argument,
+      "Could not allocate unsupported type '%s' to return register in function '%s'",
+      remill::LLVMThingToString(ret_type).c_str(),
+      function.getName().str().c_str());
+}
+
+// For X86_64_SysV, the general argument passing behavior is, try to pass the
+// arguments in registers RDI, RSI, RDX, RCX, R8, R9 from integral types and
+// XMM0 - XMM7 for float types. If the argument is a struct but can be
+// completely split over the above registers, then greedily split it over the
+// registers. Otherwise, the struct is passed entirely on the stack. If we run
+// our of registers then pass the rest of the arguments on the stack.
+llvm::Error SPARC32_C::BindParameters(
+    llvm::Function &function, bool injected_sret,
+    std::vector<ParameterDecl> &parameter_declarations) {
+  CHECK(!injected_sret)
+      << "Injected struct returns are not supported on SPARC targets";
+
+  const auto param_names = TryRecoverParamNames(function);
+  llvm::DataLayout dl(function.getParent());
+
+  // Used to keep track of which registers have been allocated
+  AllocationState alloc_param(parameter_register_constraints, arch, this);
+  alloc_param.config.type_splitter = IntegerTypeSplitter;
+
+  uint64_t stack_offset = 92;
+
+  const auto sp_reg = arch->RegisterByName("o6");
+
+  for (auto &argument : function.args()) {
+    const auto &param_name = param_names[argument.getArgNo()];
+    const auto param_type = argument.getType();
+
+    auto allocation = alloc_param.TryRegisterAllocate(*param_type);
+
+    // Try to allocate from a register. If a register is not available then
+    // allocate from the stack.
+    if (allocation) {
+      const auto prev_size = parameter_declarations.size();
+
+      for (const auto &param_decl : allocation.getValue()) {
+        parameter_declarations.emplace_back();
```

Comment:
any particular reason for using .emplace_back() and then .back() versus just making a new ParameterDecl? I figure its to keep the code generic using `auto` and type inference?

Reply:
I coded this before we adopted c++17, I think.

Reply:
But also doing an emplace back, followed by a back, lets me operate on the decl in its final position.